Packs/Base/Scripts/DBotMLFetchData/DBotMLFetchData.py

```diff
@@ -1,6 +1,1031 @@
+    text_length = len(email_body)
+    subject_length = len(email_subject)
+    number_of_lines = email_body.count('\n') + 1
+    word_subject = [w for w in email_subject_word_tokenized if any(c.isalnum() for c in w)]
+    num_of_words_subject = len(word_subject)
+    ending_dots = len([s for s in sentences if s.strip().endswith('.')])
+    ending_explanation = len([s for s in sentences if s.strip().endswith('!')])
+    ending_question = len([s for s in sentences if s.strip().endswith('?')])
+
+    return {'num_of_words': num_of_words,
+            'avg_word_length': avg_word_length,
+            'num_of_sentences': num_of_sentences,
+            'avg_sentence_length': avg_sentence_length,
+            'avg_number_of_words_in_sentence': float(num_of_words) / num_of_sentences if num_of_sentences > 0 else 0,
+            'text_length': text_length,
+            'num_of_words_subject': num_of_words_subject,
+            'subject_length': subject_length,
+            'ending_dots': ending_dots,
+            'ending_explanation': ending_explanation,
+            'ending_question': ending_question,
+            'number_of_lines': number_of_lines
+            }
+
+
+def get_characters_features(text):
+    characters_count = {}
+    for c in string.printable:
+        if c in ['[', ']', '<']:
+            continue
+        count = text.count(c)
+        if count > 0:
+            characters_count[c] = count
+    return characters_count
+
+
+def get_url_features(email_body, email_html, soup):
+    http_regex = r'http://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
+    https_regex = r'https://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+'
+    http_urls = re.findall(http_regex, email_body)
+    https_urls = re.findall(https_regex, email_body)
+    embedded_urls = [a['href'] for a in soup.findAll('a')]
+    all_urls = http_urls + https_urls + embedded_urls
+    all_urls_lengths = [len(u) for u in all_urls]
+    average_url_length = 0 if len(all_urls) == 0 else sum(all_urls_lengths) / len(all_urls)
+    min_url_length = 0 if len(all_urls_lengths) == 0 else min(all_urls_lengths)
+    max_url_length = 0 if len(all_urls_lengths) == 0 else max(all_urls_lengths)
+    shortened_urls_count = len([u for u in all_urls if any(shortened_u in u for shortened_u in SHORTENED_DOMAINS)])
+    drive_count = len([u for u in all_urls if any(drive in u for drive in DRIVE_URL_KEYWORDS)])
+    return {
+        'http_urls_count': len(http_urls),
+        'https_urls_count': len(https_urls),
+        'embedded_urls_count': len(embedded_urls),
+        'all_urls_count': len(all_urls),
+        'average_url_length': average_url_length,
+        'min_url_length': min_url_length,
+        'max_url_length': max_url_length,
+        'shortened_urls_count': shortened_urls_count,
+        'drive_count': drive_count
+    }
+
+
+def get_html_features(soup):
```

Comment:
I will also the number of URLs inside the email body that the text different than the URL itself
(like "click here" where here is a link)

Reply:
it's counted as "embedded_urls" under url_features